tests/OutputTest/OutputTest.php

```diff
@@ -0,0 +1,209 @@
+{
+    private const DATASET_DIRECTORY = __DIR__ . "/../_files/OutputTest";
+    private const FAILED_OUTPUT_DIRECTORY = __DIR__ . "/../../tmp/failed-output-tests";
+
+    private static function datasetName(SplFileInfo $file): string
+    {
+        $prefixLength = strlen(self::DATASET_DIRECTORY);
+        $path = substr($file->getPath(), $prefixLength + 1);
+        $name = $file->getBasename("." . $file->getExtension());
+        return "$path/$name";
+    }
+
+    /**
+     * @return Iterator<Dataset>
+     */
+    public static function datasets(): Iterator
+    {
+        $flags = FilesystemIterator::KEY_AS_FILENAME
+            | FilesystemIterator::CURRENT_AS_FILEINFO
+            | FilesystemIterator::SKIP_DOTS;
+        $filter = function (SplFileInfo $file) {
+            return $file->getExtension() === "html";
+        };
+        $dir = new RecursiveDirectoryIterator(self::DATASET_DIRECTORY, $flags);
+        $files = new CallbackFilterIterator(new RecursiveIteratorIterator($dir), $filter);
+
+        foreach ($files as $file) {
+            $name = self::datasetName($file);
+            yield new Dataset($name, $file);
+        }
+    }
+
+    public static function outputTestProvider(): Iterator
+    {
+        foreach (self::datasets() as $dataset) {
+            yield $dataset->name => [$dataset];
+        }
+    }
+
+    protected function setUp(): void
+    {
+        $process = new Process(["gs", "-v"]);
+        $exitCode = $process->run();
+
+        if ($exitCode === 127) {
+            $this->markTestSkipped("Output tests need Ghostscript to be installed");
+        }
+    }
+
+    /**
+     * @dataProvider outputTestProvider
+     */
+    public function testOutputMatchesReferenceRendering(Dataset $dataset): void
+    {
+        $document = $dataset->render();
+        $referenceFile = $dataset->referenceFile()->getPathname();
+        $actualOutputFile = tempnam(sys_get_temp_dir(), "dompdf_test_");
+
+        file_put_contents($actualOutputFile, $document->output());
+
+        try {
+            $this->assertOutputMatches($referenceFile, $actualOutputFile);
+        } catch (AssertionFailedError $e) {
+            $this->saveFailedOutput($dataset);
+            throw $e;
+        } finally {
+            unlink($actualOutputFile);
+        }
+    }
+
+    private function assertOutputMatches(
+        string $referenceFile,
+        string $actualOutputFile
+    ): void {
+        $command = function ($file) {
+            return [
+                "gs",
+                "-q", "-dBATCH", "-dNOPAUSE", "-sstdout=%stderr",
+                "-sDEVICE=png16m", "-dGraphicsAlphaBits=4",
+                "-sOutputFile=-", $file
+            ];
+        };
+        $process1 = new Process($command($referenceFile));
+        $process2 = new Process($command($actualOutputFile));
+
+        $process1->mustRun();
+        $process2->mustRun();
+
+        $error1 = $process1->getErrorOutput();
```

Comment:
Does this give output as `mustRun` docs says:
```
The mustRun() method is identical to run(), except that it will throw a [ProcessFailedException](https://github.com/symfony/symfony/blob/6.2/src/Symfony/Component/Process/Exception/ProcessFailedException.php) if the process couldn't be executed successfully (i.e. the process exited with a non-zero code):
```

Reply:
Yes, the message for `ProcessFailedException` contains all the relevant information, so something like:

```
Symfony\Component\Process\Exception\ProcessFailedException: The command "'gs' '-q' '-dBATCH' '-dNOPAUSE' '-sstdout=%stderr' '-sDEVICE=incorrect' '-dGraphicsAlphaBits=4' ...'" failed.

Exit Code: 1(General error)

Working directory: ...

Output:
================


Error Output:
================
Unknown device: incorrect
```

I added a short comment to the error-output handling below and changed the wording, so that it should be more obvious what is happening. Error output could be produced with e.g. the option `-dFirstPage` with a non-exisiting page, which would output something like:

```
Requested FirstPage is greater than the number of pages in the file: 2
   No pages will be processed (FirstPage > LastPage).
```

The command does not fail in that case, in contrast to specifying an incorrect value for`-sDevice`.